<a href="https://colab.research.google.com/github/MariadelMar04/Analitica-3/blob/main/Reto1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Reto 1 – Car Crash – Naive Bayes**

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
#carga de la base de datos
nxl='/content/1. BD2_CarCrash_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name='Car Crash')
XDB=XDB.dropna()
print(XDB)

            ID      County            City  Weekday  Severity  ViolCat  \
0            1   SAN DIEGO       SAN DIEGO        7         1        8   
1            2    HUMBOLDT  UNINCORPORATED        4         1        8   
2            3     VENTURA          OXNARD        2         1       12   
3            4  STANISLAUS  UNINCORPORATED        4         1        1   
4            5   MENDOCINO  UNINCORPORATED        5         1        1   
...        ...         ...             ...      ...       ...      ...   
112655  112656      SHASTA         REDDING        1         0       11   
112656  112657      SHASTA         REDDING        6         1       11   
112657  112658      SHASTA         REDDING        5         0       11   
112658  112659      SHASTA         REDDING        3         0       11   
112659  112660      SHASTA         REDDING        5         1       11   

        ClearWeather  Month CrashType  Highway  Daylight  
0                  0      1         A        0      

In [ ]:
#filtrar por ciudad
XDB=XDB[XDB['City']=='SAN FRANCISCO']
print(XDB)

#filtrar por variables
XDB=XDB[['Weekday','ViolCat','ClearWeather','Month','CrashType','Daylight','Highway']]
print(XDB)




            ID         County           City  Weekday  Severity  ViolCat  \
907        908  SAN FRANCISCO  SAN FRANCISCO        7         1        1   
1580      1581  SAN FRANCISCO  SAN FRANCISCO        2         0        9   
1581      1582  SAN FRANCISCO  SAN FRANCISCO        7         0        3   
1586      1587  SAN FRANCISCO  SAN FRANCISCO        7         0        3   
1587      1588  SAN FRANCISCO  SAN FRANCISCO        3         0        3   
...        ...            ...            ...      ...       ...      ...   
112598  112599  SAN FRANCISCO  SAN FRANCISCO        5         0       11   
112599  112600  SAN FRANCISCO  SAN FRANCISCO        5         0       11   
112600  112601  SAN FRANCISCO  SAN FRANCISCO        6         0       10   
112601  112602  SAN FRANCISCO  SAN FRANCISCO        4         0       10   
112602  112603  SAN FRANCISCO  SAN FRANCISCO        5         0       12   

        ClearWeather  Month CrashType  Highway  Daylight  
907                1      3 

In [ ]:
# Convertir Crash Type a valores numericos por orden alfabetico
XDB['CrashType']=pd.Categorical(XDB['CrashType']).codes
XDB=XDB.dropna()
print(XDB)

        Weekday  ViolCat  ClearWeather  Month  CrashType  Daylight  Highway
907           7        1             1      3          0         1        1
1580          2        9             1      2          0         0        0
1581          7        3             1      1          0         0        0
1586          7        3             1      3          0         0        0
1587          3        3             1      3          0         0        0
...         ...      ...           ...    ...        ...       ...      ...
112598        5       11             1      8          6         0        0
112599        5       11             1      3          6         1        0
112600        6       10             0      4          6         0        0
112601        4       10             1      4          6         0        0
112602        5       12             1      2          6         1        0

[2081 rows x 7 columns]


In [ ]:
#Separar variables por entrada - salida
XD=XDB[['Weekday','ViolCat','ClearWeather','Month','CrashType','Daylight']]
yd=XDB['Highway']
XDB=XDB.dropna()
print(XDB)

NameError: name 'XDB' is not defined

In [ ]:
# Implementar modelo Naive Bayes

from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,yd.values.ravel())

GaussianNB()

In [ ]:
from time import ctime
#Se obtiene la información del modelo
ct=mnb.class_count_
print("¿Como estan separados los datos?\n",ct)
print("El total de los datos es:\n",len(XD))
print("Los datos porcentualmente estan divididos:\n",ct/len(XD))

#Se obtiene la información por categoria
u=mnb.theta_
print("La media de accidentes por categoria es:\n",u)
std=mnb.var_  #Esto es varianza
std=np.sqrt(std) #Esto es desviación Estandar
print("La desviación estandar de accidentes por categoria es:\n",std)

¿Como estan separados los datos?
 [1580.  501.]
El total de los datos es:
 2081
Los datos porcentualmente estan divididos:
 [0.75925036 0.24074964]
La media de accidentes por categoria es:
 [[3.88417722 7.45696203 0.86265823 4.35189873 3.21772152 0.66265823]
 [3.9261477  4.1497006  0.79840319 6.69461078 2.27744511 0.62075848]]
La desviación estandar de accidentes por categoria es:
 [[1.95062542 3.52136997 0.34420782 2.3433645  1.87371962 0.47280262]
 [2.01603723 2.16198902 0.40119266 3.44163641 1.05765448 0.48519831]]


In [ ]:
#Identificar los limites superiores e inferiores de las categorias de clasificacion
u=mnb.theta_
print("El limite superior de los valores:\n",u+std)
print("El limite inferior de los valores:\n",u-std)

El limite superior de los valores:
 [[ 5.83480263 10.978332    1.20686605  6.69526323  5.09144114  1.13546084]
 [ 5.94218494  6.31168962  1.19959585 10.13624719  3.33509959  1.10595679]]
El limite inferior de los valores:
 [[1.9335518  3.93559205 0.51845041 2.00853424 1.3440019  0.18985561]
 [1.91011047 1.98771158 0.39721054 3.25297437 1.21979063 0.13556017]]


In [ ]:
#Indicar la categoría a la que pertenece una persona con las siguientes características [5,1,1,3,4,0]
XS = pd.DataFrame([[5, 1, 1, 3, 4, 0]], columns=XD.columns)
mnb.predict(XS)

array([0])

**Matriz de Confusión**


In [ ]:
ydp=mnb.predict(XD)  #Hazme el pronostico de los positivos y los negativos
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]

Ex=(VN+VP)/(VN+FP+FN+VP)
print("Exactitud del modelo:",Ex)

TE=(FN+FP)/(VN+FP+FN+VP)
print("Tasa de Error:",TE)

Sen=VP/(VP+FN)
print("Sensibilidad del modelo:",Sen)

Sp=VN/(VN+FP)
print("Especificidad del modelo:",Sp)

pre=VP/(VP+FP)
print("Precisión del modelo:",pre)

pren=VN/(VN+FN)
print("Precisión Negativa del modelo:",pren)

[[1365  215]
 [ 224  277]]
Exactitud del modelo: 0.7890437289764536
Tasa de Error: 0.21095627102354636
Sensibilidad del modelo: 0.5528942115768463
Especificidad del modelo: 0.8639240506329114
Precisión del modelo: 0.5630081300813008
Precisión Negativa del modelo: 0.8590308370044053


**Analisis de Resultados**



Según los resultados obtenidos, se pudo verificar en Google Colab el desempeño Naive Bayes utilizando los datos del archivo de Excel como referencia. El modelo Naive Bayes aplicado, de acuerdo a los resultados presenta una Exactitud cercana a los 80%, con una alta especifidad del 86%, lo que indica que el modelo se destaca identificando accidentes que **no** ocurrieron en autopista. Sin embargo su Sensibilidad es del 55% lo que indica que el modelo no es tan eficaz detectando accidentes que **sí** ocurrieron en autopistas.  


